aws/rust-runtime/aws-runtime/src/invocation_id.rs

```diff
@@ -9,58 +9,132 @@ use aws_smithy_runtime_api::client::interceptors::{
+    /// Call this function to receive a new [`InvocationId`] or an error explaining why one couldn't
+    /// be provided.
+    fn generate(&self) -> Result<InvocationId, BoxError>;
 }
 
-impl InvocationIdInterceptor {
-    /// Creates a new `InvocationIdInterceptor`
+/// A generator for random invocation IDs. Invocation IDs are shared across all attempts made in
+/// service of an operation and are used to help Services track requests.
+#[derive(Debug, Default)]
+pub struct RandomInvocationIdGenerator {}
+
+impl RandomInvocationIdGenerator {
+    /// Create a new `RandomInvocationIdGenerator`.
     pub fn new() -> Self {
-        Self::default()
+        Self {}
     }
 }
 
-impl Default for InvocationIdInterceptor {
-    fn default() -> Self {
+impl InvocationIdGenerator for RandomInvocationIdGenerator {
+    fn generate(&self) -> Result<InvocationId, BoxError> {
+        Ok(InvocationId::new())
+    }
+}
+
+/// A "generator" that returns [`InvocationId`]s from a predefined list.
+#[cfg(feature = "test-util")]
+#[derive(Debug)]
+pub struct InvocationIdGeneratorForTests {
+    pre_generated_ids: Arc<Mutex<Vec<InvocationId>>>,
+}
+
+#[cfg(feature = "test-util")]
+impl InvocationIdGeneratorForTests {
+    /// Given a `Vec<InvocationId>`, create a new [`InvocationIdGeneratorForTests`].
+    pub fn new(mut invocation_ids: Vec<InvocationId>) -> Self {
+        // We're going to pop ids off of the end of the list, so we need to reverse the list or else
+        // we'll be popping the ids in reverse order, confusing the poor test writer.
+        invocation_ids.reverse();
+
         Self {
-            id: InvocationId::from_uuid(),
+            pre_generated_ids: Arc::new(Mutex::new(invocation_ids)),
         }
     }
 }
 
+#[cfg(feature = "test-util")]
+impl InvocationIdGenerator for InvocationIdGeneratorForTests {
+    fn generate(&self) -> Result<InvocationId, BoxError> {
+        Ok(self
+            .pre_generated_ids
+            .lock()
+            .expect("this will never be under contention")
+            .pop()
+            .expect("testers will provide enough invocation IDs"))
+    }
+}
+
+/// This interceptor generates a UUID and attaches it to all request attempts made as part of this operation.
+#[non_exhaustive]
+#[derive(Debug, Default)]
+pub struct InvocationIdInterceptor {}
+
+impl InvocationIdInterceptor {
+    /// Creates a new `InvocationIdInterceptor`
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
 impl Interceptor for InvocationIdInterceptor {
     fn modify_before_retry_loop(
         &self,
-        context: &mut BeforeTransmitInterceptorContextMut<'_>,
-        _cfg: &mut ConfigBag,
+        _ctx: &mut BeforeTransmitInterceptorContextMut<'_>,
+        cfg: &mut ConfigBag,
     ) -> Result<(), BoxError> {
-        let headers = context.request_mut().headers_mut();
-        let id = _cfg.get::<InvocationId>().unwrap_or(&self.id);
+        let gen = cfg
+            .get::<Box<dyn InvocationIdGenerator>>()
```

Comment:
Overriding this is the exception, so I think we should check for a generator in the bag, but otherwise default to the random UUID generator (and more generally avoid an allocation for this thing in the common case).